pkgs/native_assets_builder/lib/src/utils/file.dart

```diff
@@ -50,4 +51,69 @@ extension DirectoryExtension on Directory {
     }
     return last;
   }
+
+  /// Run [callback] with this Dart process having exclusive access to this
+  /// [Directory].
+  ///
+  /// Note multiple isolates and isolate groups in the same Dart process share
+  /// locks, so these will be able to enter the exclusive section simultanously.
+  ///
+  /// The required [timeout] parameter determines how long to retry before
+  /// giving up. If the [timeout] is exceeded a [TimeoutException] is thrown.
+  // TODO(dacoharkes): Can we also make it in-process safe?
+  // TODO(dacoharkes): Where do we place this code so hook writers can reuse it?
+  // I don't really want to add it to package:native_assets_cli. A new package
+  // in dart-lang/tools?
+  Future<T> exclusive<T>(
+    Future<T> Function() callback, {
+    required Duration timeout,
+  }) async {
+    final timeoutCompleter = Completer<void>();
+    final timeoutTimer = Timer(timeout, timeoutCompleter.complete);
+    const lockFileName = '.lock';
+    final lockFile = File.fromUri(uri.resolve(lockFileName));
+    final fileSystemExceptions = <FileSystemException>[];
+    while (true) {
+      try {
+        if (!await lockFile.exists()) {
+          await lockFile.create(recursive: true);
+        }
+        final randomAccessFile = await lockFile.open(mode: FileMode.write);
+        await lockFile.writeAsString(
+          'Last acquired by ${Platform.resolvedExecutable} '
+          'running ${Platform.script} on ${DateTime.now()}.',
+        );
+        final fileLockCompleter =
+            _completer(randomAccessFile.lock(FileLock.blockingExclusive));
+        await Future.any([
+          fileLockCompleter.future,
+          timeoutCompleter.future,
+        ]);
+        if (timeoutCompleter.isCompleted) {
+          var message = 'Could not acquire the lock to ${lockFile.path}.';
+          for (final e in fileSystemExceptions) {
+            message += '\n${e.message}';
+          }
+          throw TimeoutException(message, timeout);
```

Comment:
I'm releasing the lock now. 👍

But that still means the Dart process won't terminate if it's still waiting on a lock I suppose?
We don't have a blocking exclusive lock that you can cancel.

Reply:
https://github.com/flutter/flutter/blob/80ce2b4fad8bbcbb90ec8a42cb554961f5cf6237/packages/flutter_tools/lib/src/cache.dart#L333-L356

```dart
    bool locked = false;
    bool printed = false;
    while (!locked) {
      try {
        _lock!.lockSync();
        locked = true;
      } on FileSystemException {
        if (!printed) {
          _logger.printTrace('Waiting to be able to obtain lock of Flutter binary artifacts directory: ${_lock!.path}');
          // This needs to go to stderr to avoid cluttering up stdout if a
          // parent process is collecting stdout (e.g. when calling "flutter
          // version --machine"). It's not really a "warning" though, so print it
          // in grey. Also, make sure that it isn't counted as a warning for
          // Logger.warningsAreFatal.
          _logger.printWarning(
            'Waiting for another flutter command to release the startup lock...',
            color: TerminalColor.grey,
            fatal: false,
          );
          printed = true;
        }
        await Future<void>.delayed(const Duration(milliseconds: 50));
      }
    }
```

Flutter tools uses `FileLock.exclusive` instead of `FileLock.exclusiveBlocking` and retries every 50 milliseconds.

This means that on a timeout the Dart process isn't stuck until it acquires a lock. So maybe that's a better implementation to go for.

(The flutter_tools locking has no timeout duration, but that would be easy to support as well.)

cc @mkustermann

Reply:
> We don't have a blocking exclusive lock that you can cancel.

That's because the operating system don't provide it, I suppose.

> This means that on a timeout the Dart process isn't stuck until it acquires a lock. So maybe that's a better implementation to go for.

:+1: